Key Metrics “Connect GA” CTA

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6265 

## Relevant technical choices

This PR implements the Key Metrics “Connect GA” CTA.

### Difference from the IB

If you take a look at the Figma designs, it can be noticed that the **Key Metrics Setup CTA widget** and the **Connect GA4 CTA widget** have a lot in common when it comes to the layout.

As a result, instead of following the IB in this regard, I have used the layout elements implemented in the **Key Metrics Setup CTA widget** as a part of #6210 and extracted them into common layout components to re-use them in both CTA widgets.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

aaemnnosttv commented on assets/js/googlesitekit/widgets/register-defaults.js:
```diff
@@ -233,4 +235,18 @@ export function registerDefaults( widgetsAPI ) {
+				width: [ widgetsAPI.WIDGET_WIDTHS.FULL ],
+				priority: 1,
+				wrapWidget: false,
+				modules: [ 'analytics' ],
```
This should be `analytics-4` since this is only relevant for the GA4 version of the dashboard.

Additionally, we should probably use the `isActive` property here to avoid rendering this component at all if we know it shouldn't be shown (e.g. it's dismissed or GA4 is connected).

nfmohit replied:
> This should be `analytics-4` since this is only relevant for the GA4 version of the dashboard.

Updated, thank you!

> Additionally, we should probably use the `isActive` property here to avoid rendering this component at all if we know it shouldn't be shown (e.g. it's dismissed or GA4 is connected).

Updated this to use `isActive` for three obvious cases: CTA is not dismissed, user input is completed, and GA4 is not connected. Also took into consideration the fact that these data can be `undefined` when loading.

Let me know if this looks good now, thanks!

aaemnnosttv commented on assets/js/components/KeyMetrics/KeyMetricsCTAFooter.js:
```diff
@@ -0,0 +1,49 @@
+		<Row>
+			<Cell size={ 12 }>
+				<Link onClick={ onActionClick }>
+					{ __( 'Maybe later', 'google-site-kit' ) }
```
The style doesn't quite match the design. This should use `surfaces/on-background-variant` but it's currently green.

This applies to the setup CTA too, so fixing it here should cover that one too 👍

nfmohit replied:
Fixed the colour including a few more styles, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.js:
```diff
@@ -0,0 +1,208 @@
+				actions={
+					<SpinnerButton
+						onClick={ handleCTAClick }
+						isSaving={ inProgress }
```
Please also make the button `disabled` when saving. This should happen by default IMO but it doesn't work that way currently.

nfmohit replied:
Excellent spot, updated, thanks!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.stories.js:
```diff
@@ -0,0 +1,113 @@
+					widgetSlugs: keyMetricWidgets,
+				} );
+
+				registry
+					.dispatch( CORE_WIDGETS )
+					.registerWidgetArea(
+						AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY,
+						{
+							title: 'Key metrics',
+						}
+					);
+				registry
+					.dispatch( CORE_WIDGETS )
+					.assignWidgetArea(
+						AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY,
+						CONTEXT_MAIN_DASHBOARD_KEY_METRICS
+					);
+
+				keyMetricWidgets.forEach( ( slug ) => {
+					registry.dispatch( CORE_WIDGETS ).registerWidget( slug, {
+						Component: () => <div>Hello test.</div>,
+						modules: [ 'analytics-4' ],
+					} );
+					registry
+						.dispatch( CORE_WIDGETS )
+						.assignWidget(
+							slug,
+							AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY
+						);
+				} );
+			};
```
Can we replace this with `registerDefaults` instead?

nfmohit replied:
I've created `provideWidgetRegistrations` and have used that to register the widgets into the registry. Let me know if that looks good, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+		] );
+	} );
+
+	it( 'should not render when the widget is dismissed', async () => {
+		registry
+			.dispatch( CORE_USER )
+			.receiveGetDismissedItems( [ DISMISSED_ITEM_KEY ] );
+
+		const { container, waitForRegistry } = render(
+			<ConnectGA4CTAWidget Widget={ Widget } WidgetNull={ WidgetNull } />,
+			{
+				registry,
+			}
+		);
+		await waitForRegistry();
+		expect( container ).toHaveTextContent( 'NULL' );
+	} );
```
Similar to my previous comment, we should ideally guard the widget from rendering until we know it is not dismissed to avoid invoking unnecessary side effects.

nfmohit replied:
I've moved this condition to check if the CTA is dismissed into the `isActive` parameter of the widget registration, also including returning `false` if the dismissals have not loaded. As a result, I've removed the condition from the component and subsequently removed this test case too. Let me know if that looks good, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+		expect( container ).toHaveTextContent( 'NULL' );
+	} );
+
+	it( 'should not render if GA4 is connected', async () => {
+		provideModules( registry, [
+			{
+				slug: 'analytics-4',
+				active: true,
+				connected: true,
+			},
+		] );
+
+		const { container, waitForRegistry } = render(
+			<ConnectGA4CTAWidget Widget={ Widget } WidgetNull={ WidgetNull } />,
+			{
+				registry,
+			}
+		);
+		await waitForRegistry();
+		expect( container ).toHaveTextContent( 'NULL' );
+	} );
```
This should probably also be another condition where we guard the widget from rendering from outside.

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+
+	const DISMISSED_ITEM_KEY = 'key-metrics-connect-ga4-cta-widget';
+
+	const Widget = ( { children } ) => <div>{ children }</div>;
+	const WidgetNull = () => <div>NULL</div>;
```
Why override these? I see that this has been done in a few other places but don't see why we'd not use the normal components. We're already doing that in the story, so we should be able to do the same here https://github.com/google/site-kit-wp/blob/ea139720edaec9027893f64e02c9cd5c775ed28b/assets/js/components/KeyMetrics/ConnectGA4CTAWidget.stories.js#L37-L39

nfmohit replied:
Good idea. I've used this in the tests too. Thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+describe( 'ConnectGA4CTAWidget', () => {
+	let registry;
+
+	const DISMISSED_ITEM_KEY = 'key-metrics-connect-ga4-cta-widget';
```
Let's move the definition here to a shared location to avoid duplicates.

nfmohit replied:
Moved to the datastore constants file. Thanks!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+	} );
+
+	it( 'should not render if user input is not completed', async () => {
+		global._googlesitekitUserData.isUserInputCompleted = false;
+		await registry
+			.dispatch( CORE_USER )
+			.receiveIsUserInputCompleted( false );
```
It seems unnecessary that we'd need to both set the global and receive the same value directly. We should be able to use the datastore exclusively – using the global in a test should only be done in the tests that are intended to exercise the resolver which uses it.

nfmohit replied:
This test case has been removed. However, as far as I can remember, I was encountering an error when not setting the global, which is odd and needed to be investigated further.

nfmohit replied:
This test case has been removed. However, as far as I can remember, I was encountering an error when not setting the global, which is odd and needed to be investigated further.

aaemnnosttv replied:
Thanks! The problem seems to be here in the resolver where it only returns early if user input has been completed https://github.com/google/site-kit-wp/blob/9507eaf5369ebbdac51858154b49e3fc43a9af92/assets/js/googlesitekit/datastore/user/user-info.js#L267-L269

This should be changed to return early any time the value is not `undefined`.

So even though we're providing the value directly, the resolver is still trying to get the value (and receive it again) from the global.

nfmohit replied:
Excellent catch. Understood and updated, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.js:
```diff
@@ -0,0 +1,208 @@
+	const isAnalyticsConnected = useSelect( ( select ) =>
+		select( CORE_MODULES ).isModuleConnected( 'analytics' )
+	);
+	const connectGA4URL = `${ settingsURL }#connected-services/analytics/edit`;
```
This can be combined into the selector for `settingsURL` since that isn't used anywhere else 👍

nfmohit replied:
Updated, thanks!

nfmohit replied:
Updated, thanks!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.js:
```diff
@@ -0,0 +1,208 @@
+
+			await snapshotAllStores();
+
+			navigateTo( `${ settingsURL }#connected-services/analytics/edit` );
```
Should use `connectGA4URL`, no?

nfmohit replied:
Absolutely correct, updated, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.stories.js:
```diff
@@ -0,0 +1,113 @@
+Default.storyName = 'ConnectGA4CTAWidget';
+Default.scenario = {
+	label: 'KeyMetrics/ConnectGA4CTAWidget',
+	delay: 250,
```
Is the delay necessary? Let's avoid adding it unless needed.

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.stories.js:
```diff
@@ -0,0 +1,113 @@
+		( Story ) => {
+			const setupRegistry = ( registry ) => {
+				registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );
+				global._googlesitekitUserData.isUserInputCompleted = true;
```
Similar to above, we should receive the user input state rather than mutating a global

nfmohit replied:
Removed this from the story as the sole component no longer requires user input to be completed, as that condition has been moved to the widget registration.

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+		expect( container ).toHaveTextContent( 'NULL' );
+	} );
+
+	it( 'should not render unless at least 3 analytics dependant metrics are registered', async () => {
```
Registered or set/chosen to be displayed in the widget?

nfmohit replied:
Updated, thanks!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.test.js:
```diff
@@ -0,0 +1,215 @@
+			widgetSlugs: keyMetricWidgets,
+		} );
+
+		registry
+			.dispatch( CORE_WIDGETS )
+			.registerWidgetArea( AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY, {
+				title: 'Key metrics',
+			} );
+		registry
+			.dispatch( CORE_WIDGETS )
+			.assignWidgetArea(
+				AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY,
+				CONTEXT_MAIN_DASHBOARD_KEY_METRICS
+			);
+
+		keyMetricWidgets.forEach( ( slug ) => {
+			registry.dispatch( CORE_WIDGETS ).registerWidget( slug, {
+				Component: () => <div>Hello test.</div>,
+				modules: [ 'analytics-4' ],
+			} );
+			registry
+				.dispatch( CORE_WIDGETS )
+				.assignWidget( slug, AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY );
+		} );
```
This is duplicated in a few places now. If `registerDefaults` isn't suitable here, perhaps a new utility would make sense? E.g. `provideKeyMetricsWidgetRegistrations`

nfmohit replied:
Created `provideWidgetRegistrations`, let me know if that looks good, thank you!

aaemnnosttv reviewed (changes requested):
Thanks @nfmohit – this is off to a great start. A few things to address and suggestions for you to consider.

Also, VRT is currently failing a few scenarios.

nfmohit commented:
> Thanks @nfmohit – this is off to a great start. A few things to address and suggestions for you to consider.
> 
> Also, VRT is currently failing a few scenarios.

Addressed all of them, and updated VRT scenarios. Back to you for another pass, thank you @aaemnnosttv!

aaemnnosttv commented on assets/js/googlesitekit/widgets/register-defaults.js:
```diff
@@ -233,4 +240,40 @@ export function registerDefaults( widgetsAPI ) {
 		[ AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY ]
 	);
+
+	if ( isFeatureEnabled( 'userInput' ) ) {
+		widgetsAPI.registerWidget(
+			'keyMetricsConnectGA4CTA',
+			{
+				Component: ConnectGA4CTAWidget,
+				width: [ widgetsAPI.WIDGET_WIDTHS.FULL ],
+				priority: 1,
+				wrapWidget: false,
+				modules: [ 'analytics-4' ],
```
Let's move this widget registration to the GA4 module since it's essentially an activation CTA (all widgets are registered regardless of activation/connection). This module-specific widget shouldn't be registered from a core store. Alternatively, if GA was not available as a module at all, this CTA wouldn't be registered and shouldn't be shown.

nfmohit replied:
Good call! I've moved this inside the GA4 module. Thank you!

Commit: https://github.com/google/site-kit-wp/pull/7255/commits/5cee01e5d5b6cd993bb5a5e9a5587a0d75accf14

aaemnnosttv commented on assets/js/googlesitekit/datastore/user/constants.js:
```diff
@@ -71,3 +71,6 @@ export const keyMetricsGA4Widgets = [
 	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
 ];
+
+export const KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY =
+	'key-metrics-connect-ga4-cta-widget';
```
Let's also move this with GA4 constants.

nfmohit replied:
Moved, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.js:
```diff
@@ -0,0 +1,206 @@
+const { useSelect, useDispatch } = Data;
+
+export default function ConnectGA4CTAWidget( { Widget, WidgetNull } ) {
+	const isGA4Connected = useSelect( ( select ) =>
+		select( CORE_MODULES ).isModuleConnected( 'analytics-4' )
+	);
```
This seems incorrect now that the widget will only be shown when GA4 isn't connected by the check in `isActive`. Should the `isActive` check instead be a check for module activation instead?

nfmohit replied:
This is redundant. The `isActive` check is correct. We were using this in the past to determine if only GA4 was disconnected, or the entire Analytics module was disconnected.

I've removed this and simplified the CTA handling.

Commit: https://github.com/google/site-kit-wp/pull/7255/commits/e44ad8391eb259358036823dc9ce04895b32b5ff

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.js:
```diff
@@ -0,0 +1,206 @@
+				keyMetrics.includes( slug ) && modules.includes( 'analytics-4' )
+		);
+	} );
+	const isAnalyticsActive = useSelect( ( select ) =>
+		select( CORE_MODULES ).isModuleActive( 'analytics' )
+	);
```
See above

nfmohit replied:
Addressed, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/ConnectGA4CTAWidget.js:
```diff
@@ -0,0 +1,206 @@
+
+		return `${ settingsURL }#connected-services/analytics/edit`;
+	} );
+	const isAnalyticsConnected = useSelect( ( select ) =>
+		select( CORE_MODULES ).isModuleConnected( 'analytics' )
+	);
```
Is this needed?

nfmohit replied:
It is still needed to determine if we only need to connect GA4, or setup Analytics from the ground-up. Addressed, thank you!

aaemnnosttv reviewed (commented):
Thanks @nfmohit – I think we're almost there! Just a few things left for you.

nfmohit commented on tests/js/utils.js:
```diff
@@ -420,6 +421,44 @@ export const provideKeyMetrics = ( registry, extraData = {} ) => {
+	areaTitle,
+	contextSlugs = null,
+	widgets
+) => {
```
Thank you for the kind explanation and observation.

I absolutely agree. I have created #7264, and made the current utility specific to key metrics widgets with appropriate notes. Let me know if they look good, thanks!

> Finally, just a note about the function signature that was chosen here – we generally avoid functions with more than two positional arguments because it makes it difficult to reason about what each argument is when there are many. For this reason, you'll see the `fn( a, args )` or `fn( a, options )` as conventions we use quite a bit with the difference being whether or not the second argument is required or not.

Duly noted, understood, and agreed, thank you!

aaemnnosttv commented on assets/js/modules/analytics-4/components/widgets/ConnectGA4CTAWidget.js:
```diff
@@ -109,13 +104,7 @@ export default function ConnectGA4CTAWidget( { Widget, WidgetNull } ) {
-				activateAnalytics();
-			}
-		} else if ( ! isGA4Connected ) {
+		if ( isAnalyticsConnected ) {
```
I think this should be if it's _not_ connected, no? If it were connected, we wouldn't be sending them to the connect URL, correct?

nfmohit replied:
Inside the component, we're certain that `analytics-4` is not connected (due to the widget being not active otherwise). So, if `analytics` is connected (i.e. `isAnalyticsConnected`), we're assuming that UA is connected but GA4 is not. Hence, we're sending the user to the Analytics `SettingsEdit` with the tooltip to connect GA4.

https://github.com/google/site-kit-wp/blob/a67951f57a328ac54e65d2a461ec8308dbdbb00d/assets/js/modules/analytics-4/components/widgets/ConnectGA4CTAWidget.js#L106-L118

`connectGA4URL`:
https://github.com/google/site-kit-wp/blob/a67951f57a328ac54e65d2a461ec8308dbdbb00d/assets/js/modules/analytics-4/components/widgets/ConnectGA4CTAWidget.js#L84-L90

<hr>

If `isAnalyticsConnected` is falsey, we can assume that UA is not connected as well, thus, we need to set up Analytics from the ground-up (i.e. re-auth, account selection, etc.)

Is the above logic not correct based on the AC requirements? Please let me know if I interpret them incorrectly, thanks!

aaemnnosttv replied:
Ah, I see. I don't think UA is something we need to consider here because GA4 being connected is a prerequisite to enabling KMW.

The AC says
> The banner should be rendered if more than 2 of the selected metrics are dependant on GA4 and GA4 was disconnected after the widget's setup.

The only way GA4 could be disconnected is if Analytics module was disconnected. Unlike UA, it can't be disconnected independently. GA also can't be set up without connecting GA4 so it's essentially not possible for KMW to be set up with only UA and no GA4.

nfmohit replied:
Got it, thank you for the explanation. I've updated the function and removed the unnecessary selector calls.

aaemnnosttv commented on assets/js/modules/analytics-4/components/widgets/ConnectGA4CTAWidget.js:
```diff
@@ -125,15 +114,20 @@ export default function ConnectGA4CTAWidget( { Widget, WidgetNull } ) {
+			return completeAnalyticsActivation();
+		}
+
+		activateAnalytics();
```
This call seems unreachable, since the module is either connected or active, both of which have conditions above this.

There should only be 2 code paths here, correct? active or inactive – if connected this widget shouldn't be rendered at all.

nfmohit replied:
What if the Analytics module was completely disconnected after User Input completion, in that case, it would be inactive, right? Here's a screenshot of this scenario:

![image](https://github.com/google/site-kit-wp/assets/20284937/65931601-7e19-4662-b19d-3a71147463de)

Again, please let me know if I'm misinterpreting the requirements here, thank you!

aaemnnosttv replied:
As discussed, we don't need to consider UA here 👍

aaemnnosttv commented on assets/js/modules/analytics-4/components/widgets/index.js:
```diff
@@ -26,3 +26,4 @@ export { default as TopCitiesWidget } from './TopCitiesWidget';
 export { default as TopCountriesWidget } from './TopCountriesWidget';
 export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
 export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
+export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
```
Nit: please sort 😄 There's actually a related `ConnectGA4CTATileWidget` component already so we've found the right place 👍

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/components/KeyMetrics/utils.js:
```diff
@@ -58,3 +61,41 @@ export function whenKeyMetricsWidgetVisible() {
+ * @param {Object} registry The registry to set up.
+ * @param {Object} widgets  Object containing options mapped to widget slugs.
+ */
+export const provideKeyMetricsWidgetRegistrations = ( registry, widgets ) => {
```
As a test-only utility, please move this to a module specific to tests so that we don't bundle it in production. Maybe webpack is smart enough to remove it as dead code but better to keep test sources separate regardless 👍

nfmohit replied:
Moved, let me know if it looks good, thank you!

aaemnnosttv reviewed (commented):
Thanks @nfmohit ! So close 😄

nfmohit commented:
> Thanks @nfmohit ! So close 😄

Back to you for another pass, thank you @aaemnnosttv!

aaemnnosttv reviewed (approved):
Great, thanks for your patience on this @nfmohit !
